### PR TITLE
Relax type of mimebundle data to allow for Object

### PR DESCRIPTION
**Pull request by rgbkrk**
This allows the data from a mimebundle to be an `Object`. Note that message spec v5 states that

> application/json data should be unpacked JSON data, not double-serialized as a JSON string.

Which was broadened in https://github.com/jupyter/nbformat/issues/42 and  https://github.com/jupyter/nbformat/pull/43 to allow `application/(.+\+)?json` to use raw JSON rather than double encoded payloads. This PR allows for the handling of both.

I'll follow this up with a custom mimetype for GeoJSON.

### CODE REVIEW

**Comment by sccolbert** [comment]
What is the `Map` type in this context? Is it the actual ES6 Map type (with a `get()` method), or is it a typedef for a vanilla `{}` object? 

Phosphor has some JSON typedefs which may be useful here:
https://github.com/phosphorjs/phosphor/blob/master/src/algorithm/json.ts#L10-L35

**Comment by rgbkrk** [comment]
Oh whoops, I've got a commit sitting locally. Hold on a sec.

**Comment by rgbkrk** [comment]
> Phosphor has some JSON typedefs which may be useful here

That would be so useful if we actually allowed top level JSON types in nbformat. We sadly can only support Object, in order to be backwards compatible with the previous nbformat, where these could be strings or arrays of strings that are meant to be `JSON.parse`d again.

**Comment by rgbkrk** [comment]
I'll go ahead and use `JSONObject` from Phosphor here.

**Comment by jasongrout** [comment]
> That would be so useful if we actually allowed top level JSON types in nbformat. We sadly can only support Object, in order to be backwards compatible with the previous nbformat, where these could be strings or arrays of strings that are meant to be JSON.parsed again.

At one point, the string definition had both `string` and `string[]` to mirror the multiline string definition in the disk format, but we eventually changed it to be just `string` after conversations with @minrk, who distinguished between the on-disk format, and the rest api format.

(and I forget if this source attribute was a "multiline string" or always was just a "string")

**Comment by rgbkrk** [comment]
It was string for a while followed by multiline string to help with diffing.

**Comment by minrk** [comment]
@jasongrout client-side it's always 'string'. On-disk multi-line strings are split for better storage, but nbformat hides that serialization detail from the client-side.

**Comment by jasongrout** [comment]
> @jasongrout client-side it's always 'string'. On-disk multi-line strings are split for better storage, but nbformat hides that serialization detail from the client-side.

Yep, that's why we made the change in our typings from `string[] | string` (which is what the nbformat json spec said) to just `string`.

**Comment by minrk** [comment]
I should clarify: it's always a string *if it's not JSON*. JSON is always an object from nbformat v4. @rgbkrk's PR expands the "is it JSON?" validation check from 'application/json' to the more correct value 'application/(.*+)?json'.

File: src/rendermime/index.ts [hunk 1]
```diff
@@ -6,6 +6,10 @@ import {
 } from 'phosphor/lib/core/token';
 
 import {
+  JSONObject
+} from 'phosphor/lib/algorithm/json';
```
Comment (sccolbert) [inline src/rendermime/index.ts]: We ascii order imports based on module name. This should go before `phosphor/lib/core`.

File: src/renderers/widget.ts [hunk 2]
```diff
@@ -272,7 +272,7 @@ class RenderedJavascript extends Widget {
     super();
     let s = document.createElement('script');
     s.type = options.mimetype;
-    s.textContent = options.source;
+    s.textContent = <string>options.source;
```
Comment (sccolbert) [inline src/renderers/widget.ts]: The recommended way to cast in Typescript is the `as` keyword (brackets are an old relic): `options.source as string`.

**Comment by sccolbert** [comment]
Great! Made a couple of final 💄 comments.

File: src/renderers/widget.ts (hunk 2 above)
Comment (rgbkrk) [inline src/renderers/widget.ts]: Awesome, thanks. Yeah my knowledge here was a bit outdated.

**Comment by rgbkrk** [comment]
All set.

**Comment by jasongrout** [comment]
Please don't forget lines 247 and 312 (you can search for `options.source`...)

**Comment by jasongrout** [comment]
This looks good to me now. I think perhaps we should distinguish between string source and json source in the future, perhaps with a `sourceType` field with values `'string'` or `'json'` (which scales to future `+xml` and other formats...

**Comment by sccolbert** [comment]
Thanks!

**Pull request merged**